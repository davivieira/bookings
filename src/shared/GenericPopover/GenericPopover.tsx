/* eslint-disable @typescript-eslint/no-explicit-any */
import * as React from 'react';
import Popover from '@mui/material/Popover';

type TriggerButtonProps = {
  onClick: (event: React.MouseEvent<HTMLButtonElement>) => void;
};

type GenericPopoverProps = {
  children: React.ReactNode;
  TriggerButton: React.ComponentType<TriggerButtonProps>;
};

function GenericPopover({ children, TriggerButton }: GenericPopoverProps) {
  const [anchorEl, setAnchorEl] = React.useState<HTMLButtonElement | null>(null);

  const handleClick = (event: React.MouseEvent<HTMLButtonElement>) => {
    setAnchorEl(event.currentTarget);
  };

  const handleClose = () => {
    setAnchorEl(null);
  };

  const open = Boolean(anchorEl);
  const id = open ? 'simple-popover' : undefined;

  return (
    <>
      <TriggerButton onClick={handleClick} />
      <Popover
        id={id}
        open={open}
        anchorEl={anchorEl}
        onClose={handleClose}
        anchorOrigin={{
          vertical: 'bottom',
          horizontal: 'left',
        }}
      >
        {React.isValidElement(children) &&
          React.cloneElement(children as React.ReactElement<any>, { handleClose })}
      </Popover>
    </>
  );
}

export default GenericPopover;