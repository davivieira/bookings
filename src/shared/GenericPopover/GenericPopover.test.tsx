import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';
import GenericPopover from './GenericPopover';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
jest.mock('@mui/material/Popover', () => (props: any) => (
  <div id={props.id} data-testid="popover">
    {props.open ? props.children : null}
  </div>
));

const TriggerButton: React.FC<{ onClick: (event: React.MouseEvent<HTMLButtonElement>) => void }> = ({ onClick }) => (
  <button onClick={onClick}>Open Popover</button>
);

const ChildComponent: React.FC<{ handleClose: () => void }> = ({ handleClose }) => (
  <div>
    <button onClick={handleClose}>Close Popover</button>
    <div>Popover Content</div>
  </div>
);

describe('GenericPopover', () => {

  it('matches snapshot', () => {
    const { asFragment } = render(
      <GenericPopover TriggerButton={TriggerButton}>
        <div>Popover Content</div>
      </GenericPopover>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders the TriggerButton', () => {
    render(
      <GenericPopover TriggerButton={TriggerButton}>
        <div>Popover Content</div>
      </GenericPopover>
    );

    expect(screen.getByText('Open Popover')).toBeInTheDocument();
  });

  it('opens the popover when TriggerButton is clicked', async () => {
    const user = userEvent.setup();

    render(
      <GenericPopover TriggerButton={TriggerButton}>
        <div>Popover Content</div>
      </GenericPopover>
    );

    await user.click(screen.getByText('Open Popover'));

    expect(screen.getByTestId('popover')).toBeInTheDocument();
    expect(screen.getByText('Popover Content')).toBeInTheDocument();
  });

  it('closes the popover when handleClose is called', async () => {
    const user = userEvent.setup();

    render(
      <GenericPopover TriggerButton={TriggerButton}>
        <ChildComponent handleClose={() => null} />
      </GenericPopover>
    );

    await user.click(screen.getByText('Open Popover'));
    expect(screen.getByText('Popover Content')).toBeInTheDocument();

    await user.click(screen.getByText('Close Popover'));
    expect(screen.queryByText('Popover Content')).not.toBeInTheDocument();
  });

  it('passes handleClose to children', async () => {
    const user = userEvent.setup();

    render(
      <GenericPopover TriggerButton={TriggerButton}>
        <ChildComponent handleClose={() => null} />
      </GenericPopover>
    );

    await user.click(screen.getByText('Open Popover'));

    const closeButton = screen.getByText('Close Popover');
    expect(closeButton).toBeInTheDocument();

    await user.click(closeButton);
    expect(screen.queryByText('Popover Content')).not.toBeInTheDocument();
  });
});
