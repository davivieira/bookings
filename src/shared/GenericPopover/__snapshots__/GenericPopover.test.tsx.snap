// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GenericPopover matches snapshot 1`] = `
<DocumentFragment>
  <button>
    Open Popover
  </button>
  <div
    data-testid="popover"
  />
</DocumentFragment>
`;
